<html>
<head>
  <meta http-equiv="CONTENT-TYPE" content="text/html; charset=iso-8859-15">
  <link rel="stylesheet" href="../styles.css" type="text/css">
  <title>oPOSSUM 3.0 Help</title>
</head>
<style type="text/css">
	body {
		margin:20px;
	}
	.contents_level1 {
		margin:5px;
	}
	.contents_level2 {
		margin:25px;
	}
	.contents_level3 {
		margin:45px;
	}
</style>

<body class="text">

<a name="top"></a>

  <h1><a name="ssa_help"></a>oPOSSUM anchored Combination Site Analysis (aCSA)</h1>
  <p>
  <b><center>Identification of pairs of over-represented transcription factor binding sites in sets of genes or sequences</center></b>
  </p>
  
  <a href="#analysis_input"><h2>Analysis Input</h2></a>
  
  <a class="contents_level1" href="#select_target">Entering Input Data Sets</a><br />
  <a class="contents_level2" href="#select_background">Entering Background Data Sets</a><br />
  
  <a class="contents_level1" href="#select_tfbs">Selecting transcription factor binding site profiles</a><br />
  <a class="contents_level2" href="#jaspar_core">Select an anchoring TF</a><br />
  <a class="contents_level2" href="#jaspar_pbm">Select TFs to find in combination with the anchoring TF</a><br />
  
  <a class="contents_level1" href="search_tfbs">TFBS Search Parameters</a><br />
  <a class="contents_level2" href="#conservation">Conservation level</a><br />
  <a class="contents_level2" href="#threshold">Matrix match threshold</a><br />
  <a class="contents_level2" href="#up_down_seq">Upstream / downstream sequence</a><br />
  <a class="contents_level2" href="#num_results">Number of results to return</a><br />
  <a class="contents_level2" href="#sort_by">Sorting results</a><br />
  
  <a href="#analysis_results"><h2>Analysis Results</h2></a>
  
  <a class="contents_level1" href="#selected_parameters">Selected Parameters</a><br />
  <a class="contents_level2" href="#num_submitted_genes">Number of submitted genes/sequences</a><br />
  <a class="contents_level2" href="#selected_tfbs">Selected TFBS</a><br />
  <a class="contents_level2" href="#search_parameters">Search parameters</a><br />
  <a class="contents_level2" href="#gc_comp">GC composition</a><br />
  <a class="contents_level2" href="sorted_by">Results returned</a><br />
  
  <a class="contents_level1" href="#opossum_results">Analysis Results Summary Table</a><br />
  <a class="contents_level1" href="#tfbs_hits">TFBS Hits Table</a><br />
   
  <p><hr noshade><br><a href="#top"><img src="../images/arrow-up.gif" align="right" /></a></p>  
  
  <h2><a name="analysis_input"></a>Analysis Input</h2>

  <p><br></p>
  
  <h3><a name="select_target"></a>Entering Target Data Sets</h3>    
  <p>
    Specify your list of target genes or sequences by either pasting them into
	the text box or using the file upload option.
  </p>
  <p>
	For gene-based analysis, the gene list should preferably be formatted with one ID per line but may also be separated by spaces, commas, semi-colons, colons, newlines or some combination thereof.
	  
	For testing purposes, pressing the <b>Use sample genes</b> button will paste
	a pre-defined set of test genes into the text area.
	
	Use the <b>Clear</b> button to clear the text area.

  You need to choose the type of gene ID from the drop down menu. The gene ID
	types available are:<br>
	<ul>
	  <li>Ensembl Gene ID
	  <li>EntrezGene ID
	  <li>RefSeq DNA/Protein ID
	  <li>EMBL (GenBank) ID
	  <li>Protein (GenBank) ID
	  <li>Uniprot/TrEMBL Acc.
	  <li>Uniprot/SwissProt Acc.
	  <li>Unigene ID
	</ul>
	Depending on the species chosen, other species-specific database IDs are
	available:
	<ul>
	  <li>HGNC Symbol (Human)
	  <li>MGI Symbol (Mouse)
	  <li>FlyBase Gene ID (Fly)
	  <li>WormBase Gene Public Name (Worm)
	  <li>WormBase Gene ID (Worm)
	  <li>SGD (Yeast)
	</ul>
	Please note that while oPOSSUM attempts to use many different ways
	of identifying genes, the difficulty in mapping between different gene
	identification schemes means that not every gene in oPOSSUM is represented in
	every scheme. Every gene in oPOSSUM has an Ensembl gene ID, so this is the
	recommended gene ID type.
  </p>
  <p>
	For sequence-based analysis, fasta-formatted sequences can be pasted into the
	text area or uploaded as a file. If you need only have genomic coordinates for
	your sequences, please use the <a href="http://main.g2.bx.psu.edu/">Galaxy
	service</a> from the UCSC.
  </p>

  <p><hr noshade><br><a href="#top"><img src="../images/arrow-up.gif" align="right" /></a></p>

  <h3><a name="select_background"></a>Entering Background Data Sets</h3>
  <p>
	For both gene- and sequence-based systems, the same restrictions apply
	as given in the Entering Target Data Sets section.
  </p>
  <p>
  	For gene-based oPOSSUM systems, the full background data sets were compiled by
	selecting genes from Ensembl which were considered to be well known, i.e. the
	"KNOWN" attribute was true and the genes had a gene symbol assigned (HGNC
	symbol for human, MGI for mouse etc.). Transcription start sites (TSSs) were
	assigned for each gene based on the core Ensembl gene TSS annotations.
	You can paste in the list of gene ID's to be used as the background data set, or
	upload a file containing the ID's. Alternatively, you can specify the number of
	random genes to be picked as the background data set, where the maximum number
	allowed is the number of genes annotated in the oPOSSUM system.
  </p>
  <p>
	The default setting for gene-based analyses is to use all genes in the oPOSSUM
	database as your background.  Alternatively, you can specify the number of genes
	to be picked randomly to be used as the background (default=5000) or provide
	your own custom background gene list. Choosing randomly selected background
	genes can reduce the required computation time compared to choosing the entire
	gene set if the number chosen is sufficiently larger than your foreground gene
	set. If you are performing multiple analyses, and you wish to compare the results
	between analyses, we recommend you use "all" or supply your own background so
	that the background is stable across analyses.
  </p>

  <p>
	For sequence-based systems, you can either choose a pre-defined background
	data set or supply your own. When choosing a background data set, it is
  important to ascertain that its GC composition closely matches that of the
  target data set. Ideally, the average lengths of the sequences used in the
  background set should also closely match the lengths used in the target set.
  The pre-defined data sets include:
	<ul>
	  <li>Mouse fibroblast cells
	  <li>Mouse liver cells
	  <li>Mouse ES cells
	  <li>Mouse bone marrow cells
	  <li>Mouse mixed cell lines (a combination of the cell lines above)
	</ul>
  </p>
  <p>
	For sequence-based analyses, the default setting for the background selection
	is to provide your own set of control sequences.  Alternatively, one can choose
	to use one of the provided backgrounds; however, these are currently limited
	to mouse sequences, and as each background set has a specific nucleotide
	composition (indicated in the background name), the available background sets
	may not necessarily be good matches for your foreground sequences.  It is
	preferable to supply your own control sequences as you can control for the
	nucleotide composition of your sequence sets. There is an automatic background
	sequence generator that is currently undergoing beta testing (provided below
	the custom background box), which you can use to extract a background with a
	nucleotide composition comparable to your foreground sequences. When choosing
	the background, you should ensure that you have the same number (or greater)
	of sequences in your background set as in the foreground set.  
  </p>
  
  <p><hr noshade><br><a href="#top"><img src="../images/arrow-up.gif" align="right" /></a></p>

  <h3><a name="select_tfbs"></a>
    Select transcription factor binding site profiles
  </h3>
    
	<p>The matrices included in oPOSSUM-3 were obtained from the 2010 release of
      <a href="http://jaspar.genereg.net" target="_blank">JASPAR</a> database.
      For gene-based analysis, you are currently restricted to JASPAR CORE profiles
	  for both the anchoring TFs and the TFs to be searched in combination.
	  The actual profiles that can be selected vary according to the species selected.
      For sequence-based analysis, you have the option of pasting in or uploading your
	  custom profiles.
	</p>
    <p>
	  The default setting is to choose "all profiles" belonging to the selected
	  taxonomic group in the JASPAR CORE collection with a minimum specificity of
	  8 bits. The specificity as bits refers to the Information Content (IC) of
	  the TF matrices. The IC, loosely speaking, combines the length and the complexity
	  of the motif into a value that describes what we know collectively about the
	  sequences a TF recognizes. The minimum specificity of 8 bits was chosen because
	  TF matrices with IC of less than 8 bits are relatively uninformative for scoring
	  DNA sequence with. If you choose to threshold matrices with a higher specificity,
	  you are limiting your analyses to matrices with stronger, more selective patterns.
	  In general, it is not necessary to use a threshold higher than 8 bits unless
	  you have specific requirements, as you may later filter your final results
	  by specificity (the IC column of the results file).
	</p>
    
	<h4><a name="anchor_tf">Select an anchoring TF</h4>
    <p>
	Select your anchoring TF from the pull down list of JASPAR CORE profiles.
	Species-specific restrictions may apply.</p>  

    <h4><a name="combine_tfs"></a>Select TFs to find in combination with the anchoring TF</h4>
    <p>To select multiple TFs, use Shift-click and/or CTRL-click (Command-click for Mac).</p>
    
	<q></q>
	</p>
  
  <p><hr noshade><br><a href="#top"><img src="../images/arrow-up.gif" align="right" /></a></p>

  <h3><a name="search_tfbs"></a>TFBS Search Parameters</h3>
  
	<h4><a name="conservation"></a>Conservation level</h4>
	<p>
	  For gene-based systems, conservation was determined by using phastCons scores
	  from the UCSC Genome Browser scoring above some minimum threshold and merging
	  these into conserved regions of minimum length 20 bp.
	</p>
	<p>
	  The conserved regions that fell within 10,000 nucleotides (species dependent) upstream and downstream of the predicted
	  transcription start site (TSS) were then scanned for binding sites using position weight matrice (PWM) models 
	  of transcription factor (TF) binding specificity from the <a href="http://jaspar.genereg.net/" target="_blank">JASPAR</a> 
	  database. These TF sites were stored in the oPOSSUM database and comprise the background set for the default analysis.
	  The background set was pre-computed with three levels of conservation filter. 
	  <ul>
		<li>Level 1 corresponds to non-coding conserved regions with a minimum percent identity of 60%</li>
		<li>Level 2 corresponds to non-coding conserved regions with a minimum percent identity of 65%</li>
		<li>Level 3 corresponds to non-coding conserved regions with a minimum percent identity of 70%</li>
	  </ul>
	  The default conservation level is 0.7 (level 3). If this is too restrictive
	  for your analysis, you can lower the conservation level to 0.6 (level 1). 
	</p>
	<p>No conservation filters are used for yeast oPOSSUM and sequence-based systems.</p>
	<p><hr noshade><br><a href="#top"><img src="../images/arrow-up.gif" align="right" /></a></p>
	
	<h4><a name="threshold"></a>Matrix score threshold</h4>
	  <p>
		  TF sites are scanned by sliding the corresponding postion weight matrix (PWM) along the sequence 
		  and scoring it at each position. The threshold is the minimum relative score used to report the 
		  position as a putative binding site. The background set was computed using a minimum threshold of 75%.
	  <ul>
		<li>Level 1 corresponds to a minimum relative score of 75%</li>
		<li>Level 2 corresponds to a minimum relative score of 80%</li>
		<li>Level 3 corresponds to a minimum relative score of 85%</li>
	  </ul>
	  </p>
    <p>
      The relative score is computed from the raw matrix score as:<br>
      rel_score = (site_score - min_matrix_score) / (max_matrix_score - min_matrix_score)
	</p>
	<p>
	  The default threshold is 80%, which is a commonly used threshold for TFBS
	  analyses using PWMs. If you have prior knowledge of which TFs are of interest
	  for your analyses and what their properties are, you may change this threshold
	  based on that knowledge (between 75% to 100%). For instance, if the matrix
	  for a TF of interest has a low IC, then you will want to use a higher threshold,
	  whereas for a TF with a high IC, you might try using a lower threshold. A
	  threshold of 80% or 85% will generally provide you with satisfactory results,
	  but if you are uncertain, we recommend multiple analyses with various thresholds.
	</p>
	  <p><hr noshade><br><a href="#top"><img src="../images/arrow-up.gif" align="right" /></a></p>
   
	 <h4><a name="threshold"></a>Maximum inter-binding distance</h4>
	  <p>
		  TFs that are located within the specified maximum inter-binding distance
		  from a predicted anchoring TFBS are counted for over-representation analysis.
		  The distance can range up to 250 bp. The default distance is 100 bp. These
		  values were chosen as a reasonable boundary for including TFs that may
		  act together for a common regulatory function.  If you are uncertain of
		  what to expect in terms of distance, perform three analyses with three
		  different distances e.g. 50, 150, and 250 bp.
	  </p>
	<p><hr noshade><br><a href="#top"><img src="../images/arrow-up.gif" align="right" /></a></p>
   
	<h4><a name="up_down_seq"></a>Upstream / Downstream sequence</h4>
	  <p class="text">
		  For gene-based oPOSSUM analyses, this refers to the size of the region
		  around the transcription start sites (TSS) which was analysed for TF
		  binding sites. The maximum amount of upstream / downstream region used
		  to pre-compute the background is species dependent, as follows.
	  </p>
 <table class="sortable" cellpadding="2" cellspacing="0" width="400">
    <tr class="table_title">
    <th>Species</th><th>Max. Upstream (bp)</th><th>Max. Downstream (bp)</th>
    </tr>
    <tr class="table_text" align="right"><td>Human, Mouse</td><td>10,000</td><td>10,000</td></tr>
    <tr class="table_text" align="right"><td>Fly</td><td>3,000</td><td>3,000</td></tr>
    <tr class="table_text" align="right"><td>Worm</td><td>1,500</td><td>1,500</td></tr>
    <tr class="table_text" align="right"><td>Yeast</td><td>1,000</td><td>Annotated 3' end of the gene</td></tr>
  </table>

  <p>
   The TFBS counts within the search regions were precomputed for various levels
   of upstream / downstream sequence. The levels for the metazoans are:
  </p>
  <table class="sortable" cellpadding="2" cellspacing="0" width="400">
    <tr class="table_title">
	  <th></th><th colspan=2>Vertebrates</th><th colspan=2>Insects</th><th colspan=2>Nematodes</th>
	</tr>
	<tr>
    <th>Search Region Level</th>
	  <th>Upstream (bp)</th><th>Downstream (bp)</th>
	  <th>Upstream (bp)</th><th>Downstream (bp)</th>
	  <th>Upstream (bp)</th><th>Downstream (bp)</th>
    </tr>
    <tr class="table_text" align="center"><td>1</td>
	  <td>10,000</td><td>10,000</td><td>3,000</td><td>3,000</td><td>1,500</td><td>1,500</td>
	</tr>
    <tr class="table_text" align="center"><td>2</td>
	  <td>10,000</td><td>5,000</td><td>3,000</td><td>2,000</td><td>1,000</td><td>1,000</td>
	</tr>
    <tr class="table_text" align="center"><td>3</td>
	  <td>5,000</td><td>5,000</td><td>2,000</td><td>2,000</td><td>1,000</td><td>500</td>
	</tr>
    <tr class="table_text" align="center"><td>4</td>
	  <td>5,000</td><td>2,000</td><td>2,000</td><td>1,000</td><td>500</td><td>500</td>
	</tr>
    <tr class="table_text" align="center"><td>5</td>
	  <td>5,000</td><td>2,000</td><td>2,000</td><td>1,000</td><td>500</td><td>500</td>
	</tr>
    <tr class="table_text" align="center"><td>6</td>
	  <td>5,000</td><td>2,000</td><td>2,000</td><td>1,000</td><td>500</td><td>500</td>
	</tr>
  </table>
  <p>
	Using one of the pre-defined search region levels will result in much faster
	computation than using custom search region lengths, as the oPOSSUM system can
	take advantage of pre-computed values stored in the database. The default levels
	were reasonable in terms of regulatory region lengths for the individual organisms.
	Alternatively, you can choose other pre-computed regions from the drop down
	list or specify your own custom region.  As the regions in the drop-down list
	are pre-computed, the analysis will complete relatively quickly. Custom regions
	are not pre-computed and thus take longer to analyse; if you don't remember
	to adhere to the max upstream and max downstream constraints, the system will
	limit the region for you.  
  </p>
  <p>
	We recommend that you perform several analyses, varying the selected search region
	level for each analyses.  Using mouse as an example, you might start with the default
	of 5,000/5,000 bp, and then examine how	the over-representaiton analyses change
	between shorter and longer search regions by comparing the results from
	2,000/2,000 bp and 10,000/10,000 bp for another.
  </p>
  
  <p>
    <b>Notes:</b><br>
    <ul>
    <li>In cases where the specified amount of downstream sequence extends beyond the annotated 3' end of a gene, the amount of downstream sequence searched is truncated at the annotated 3' end.
    <li>In cases of multiple TSSs for a given gene, search regions are applied to each TSS and, if overlapping, combined.
    </ul>
  </p>

	<h4><a name="sort_by"></a>Number of results to display:</h4>

	<p class="text">You can specify the number of results to be returned. The default
	is to return all results, but you can specify to return only the top 5, 10 or
	20 results. You can also specify to return all results which score above a given
	Z-score and Fisher score threshold. In cases where anything other than the "All"
	option is specified, the top scoring results are based on which scoring method
	was chosen in the "Sort results by" section. Also, in those cases, those results
	that are not returned, are lost, even those results that scored highly in the
	other scoring measure.
	</p>
	<p>
	  We recommend using the default of "all results", as selecting less than "all"
	  does not increase the speed of the analysis. Once you have the results file,
	  you can manually filter the scores yourself.
	</p>

	<h4><a name="sort_by"></a>Sort results by:</h4>
	<p class="text">Results can be sorted by either Z-score or Fisher score. The
	sort order lists the top scoring over-represented TFBS at the top of the list.
	The sort order is not permanent, and can be changed once the analysis is complete
	and the results page is displayed.</p>
  
	<h4><a name="email"></a>Enter your email address:</h4>
	<p class="text">
	  Depending on parameters selected, the aCSA process make take a considerable amount of time (several hours) to complete. Users are notified by email of the results when the analysis is finished. Please be assured that your e-mail address is used solely for this purpose.
	</p>
  
  <p><hr noshade><br><a href="#top"><img src="../images/arrow-up.gif" align="right" /></a></p>

  <h2><a name="analysis_results"></a>Analysis Results</a></h2>  
	<p>
	The results are returned in a table format, preceded by a summary of the
	input parameters and the GC composition of the target and background
	sequences used in the analysis. The results can be downloaded in a tab delimited
	file format (link is provided at the bottom of the table). Those results that
	do not have any hits in the background genes/sequences are flagged with a
	warning.
	</p>
	
    <h3><a name="selected_parameters"></a>Selected Parameters</h3>
	
	  <a name="user_seq_files"></a><h4>Sequence information (sequence-based analysis)</h4>
	  <p>For sequence-based analysis, the following information on the analyzed
	  sequences are displayed:
	  <ul>
		<li>If the user supplied target and/or background sequence files, the file
	  names are displayed. If one of the pre-defined background sequence sets is
	  used, its name is displayed.
	    <li>The number of target and background sequences
		<li>Total lengths of target and background sequences
	  </ul>
	  
	  <a name="num_submitted_genes"></a><h4>Number of submitted genes (gene-based analysis)</h4>
	  oPOSSUM systems are built based on Ensembl IDs. If external IDs are used for
	  gene selection, oPOSSUM may not be able to map all IDs in an one-to-one manner.
	  All calculations are based on the unique oPOSSUM genes found and their associated
	  non-coding regions.
	  <ul>
		<li>Submitted for analysis: number of user-supplied target gene IDs
		<li>Included in analysis: number of user-supplied IDs that oPOSSUM successfully
	    mapped to its internal database of genes
		<li>Missing from analysis: number of user-supplied IDs that oPOSSUM could not map
		to its internal database of genes
		<li>Unique oPOSSUM genes in analysis: number of unique genes identified by oPOSSUM
		from the user-supplied IDs (some user supplied IDs may map to a single oPOSSUM gene or vice versa)
	  </ul>

      <a name="selected_tfbs"></a><h4>TFBS profile summary</h4>
	  <p>TFBS profile matrix source: JASPAR CORE profiles / User-supplied(sequence-based analysis only)</p>
	  <ul>
		<li>Anchoring TF: name of the anchoring TF
		<li>If minimum IC was specified, taxonomic group and the IC are displayed.
		<li>If subset of the available TFs were selected, their names are listed.
	  </ul>

	  <a name="search_parameters"></a><h4>Search parameters</h4>
	  <p>For gene-based analysis, the conservation level and upstream/downstream sequence lengths
	  chosen are summarized. For both the gene-based and sequence-based analyses, the matrix score
	  threshold is shown.</p>

	  <a name="gc_comp"></a><h4>GC composition</h4>
	  <p>For both the target and background sets, the GC compositions of the sequences that were searched
	  for motifs are shown. GC content less than 0.45 or greater than 0.55 is flagged red. For
	  sequence-based analysis, it is important to make sure that both the target and background
	  sequence sets to have similar levels of GC content, as sequence composition can affect
	  TFBS search results. Please refer to the manuscript for more details.</p>
	  
      <a name="sorted_by"></a><h4>Results returned</h4>
	  <p>How the results are sorted in the summary table.</p>
	  
    <p><hr noshade><br><a href="#top"><img src="../images/arrow-up.gif" align="right" /></a></p>
	  
    <h3><a name="genes_analyzed"></a>Genes Included in Analysis (gene-based analysis)</h3>
      <p>
      This section lists the set of target and background genes that were entered by the user.
	  The background gene list is shown only if you pasted in or uploaded a gene set to be used
	  as the background (i.e. not randomly chosen). It is broken down into the sections
	  <a href="#genes_analyzed">Analyzed</a> and 
      <a href="#genes_excluded">Excluded</a>. Please see description under 
      <a href="#target_genes">Target Genes</a> of what is meant by 'Analyzed' and 'Excluded' 
      and how this may impact the results of the analysis.

      <h4><a name="genes_analyzed"></a>Analyzed</h4>
      <p>
	  This is the sub-set of genes entered by the user which were found within 
	  the oPOSSUM database, and therefore included in the analysis.
	  </p>
      <h4><a name="genes_excluded"></a>Excluded</h4>
	  <p class="text">
	  This is the sub-set of genes entered by the user which were not found within 
	  the oPOSSUM database, and therefore had to be excluded from the analysis.
	  </p>

  <p><hr noshade><br><a href="#top"><img src="../images/arrow-up.gif" align="right" /></a></p>

  <h2><a name="opossum_results"></a>Analysis Results Summary Table</h2>
	<p>For a general explanation of what the oPOSSUM analysis results mean, please
	refer to the main <a href="../help.html#results" name="results">help page</a>.</p>
	<p>
      This table contains the results of both the Z-score and Fisher analyses. The results are 
      ordered by Z-score from most to least significant (higher to lower z-score). Those columns
	  that the table can be sorted by has double arrow icons in the header. The columns 
      of the table are as follows:
    </p>
    
	<h4><a name="tf_name"></a>TF</h4>
	<p>The name of the transcription factor.</p>
    
	<h4><a name="jaspar_id"></a>JASPAR ID</h4>
	<p>The JASPAR ID of the transcription factor. Link is provided to the JASPAR summary page for the TF.</p>
	
	<h4><a name="tf_class"></a>Class</h4>
    <p>The class of TFs to which this TF belongs.</p>
	
    <h4><a name="tf_family"></a>Family</h4>
	<p>The family of TFs to which this TF belongs.</p>
	
	<h4><a name="tax_group"></a>Tax group</h4>
	<p>The taxonomic supergroup to which this TF belongs.</p>
    
	<h4><a name="tf_ic"></a>IC</h4>
	<p>
	The specificity or information content (IC) of this TFBS profile's position weight matrix. 'Extreme' IC values are flagged as this may affect the results for these profiles. A very low IC matrix will be found more frequently by random chance, i.e. more false positives. A very high IC matrix will be found infrequently such that it's possible that just a few extra or fewer hits detected in the foreground or background will more greatly affect the Fisher and Z-scores. Currently those IC values less than 9 or greater than 19 are flagged.
	</p>
	
	<h4><a name="tf_gc"></a>GC Content</h4>
	<p>The GC content of the this TFBS profile. Values less than 0.33 or greater than 0.66 are flagged red.
	
    <h4><a name="t_gene_hits"></a>Target gene/sequence hits</h4>
	<p>
	The number of genes/sequences in the included target set for which this TFBS
	was predicted within the searched regions. A link is provided to the
	<a href="#tfbs_hits">TFBS Hits</a> summary
	table that lists the actual genomic locations of the hits.
	</p>
    
	<h4><a name="t_gene_non_hits"></a>Target gene/sequence non-hits</h4>
	<p>
	The number of genes/sequences in the included target set for which this TFBS
	was NOT predicted within the searched regions.
	</p>
	
    <h4><a name="bg_gene_hits"></a>Background gene/sequence hits</h4>
	<p>
	The number of genes/sequences in the background set for which this TFBS was
	predicted within the searched regions. Results with 0 background hits are
	flagged with a warning.
	</p>
      <h4><a name="bg_gene_non_hits"></a>Background gene/sequence non-hits</h4>
	<p>
	The number of genes/sequences in the background set for which this TFBS was
	NOT predicted within the searched regions.
	</p>
	<h4><a name="bg_tfbs_hits"></a>Background TFBS hits</h4>
	<p>
	The number of times this TFBS was detected within the searched regions 
	of the background set of genes/sequences. Results with 0 background hits are
	flagged with a warning.
	</p>
	
    <h4><a name="t_tfbs_hits"></a>Target TFBS hits</h4>
	<p>
	The number of times this TFBS was detected within the searched regions 
	of the target set of genes/sequences. A link is provided to the
	<a href="#tfbs_hits">TFBS Hits</a> summary
	table that lists the actual genomic locations of the hits.
	</p>
    
	<h4><a name="bg_tfbs_rate"></a>Background TFBS rate</h4>
	<p>
	The rate of occurrence of this TFBS within the searched regions of the background 
	set of genes/sequences. The rate is equal to the number of times the site was
	predicted (background hits) multiplied by the width of the TFBS profile, divided
	by the total number of nucleotides in the searched regions of the background set.
	</p>
    
	<h4><a name="t_tfbs_rate"></a>Target TFBS rate</h4>
	<p>
	The rate of occurrence of this TFBS within the searched regions of the included 
	target set of genes/sequences. The rate is equal to the number of times the site
	was predicted (target hits) multiplied by the width of the TFBS profile, divided
	by the total number of nucleotides in the searched regions of the included target set.
	</p>
    
	<h4><a name="zscore"></a>Z-score</h4>
	<p>
	The likelihood that the number of TFBS nucleotides detected for the included target
	genes/sequences	is significant as compared with the number of TFBS nucleotides detected
	for the background set. Z-score is expressed in units of magnitude of the standard deviation.
	</p>
    
	<h4><a name="fisher_score"></a>Fisher score</h4>
	<p>
	The probability that the number of hits vs. non-hits for the included target genes/sequences 
	could have occured by random chance based on the hits vs. non-hits for the background set.
	Negative natural logarithm of the probabilities are returned as the scores.
	</p>

	<p><hr noshade><br><a href="#top"><img src="../images/arrow-up.gif" align="right" /></a></p>

	<h3><a name="tfbs_hits"></a>TFBS Hits Summary Table</h3>
      
	  <p>This table contains the gene and promoter information where the TFBS prediction is found,
	  along with individual TFBS hit locations.</p>
	  <p><a href="help_tfbs_hits_gene_anchored.html">gene-based analysis</a></p>
	  <p><a href="help_tfbs_hits_seq_anchored.html">sequence-based analysis</a></p>
	  
	  
