<h3>Selected Parameters</h3>

<table border="0" cellpadding="2" cellspacing="0">
  <tr>
    <td class="table_title">Job ID:</td>
    <td class="table_text">[%job_id%]</td>
  </tr>

  [%IF user_seq_file%]
    <tr>
      <td class="table_title">Target sequence file:</td>
      <td class="table_text">[%user_seq_file%]</td>
    </tr>
  [%END%]
  
  [%IF user_peak_pos_file%]
    <tr>
      <td class="table_title">Target sequence MCP position file:</td>
      <td class="table_text">[%user_peak_pos_file%]</td>
    </tr>
  [%END%]

  [%IF user_bg_seq_file%]
    <tr>
      <td class="table_title">Background sequence file:</td>
      <td class="table_text">[%user_bg_seq_file%]</td>
    </tr>
  [%ELSIF bg_seq_set_name%]
    <tr>
      <td class="table_title">Background sequence set:</td>
      <td class="table_text">[%bg_seq_set_name%]</td>
    </tr>
  [%END%]
  
  [%IF user_bg_peak_pos_file%]
    <tr>
      <td class="table_title">Background sequence MCP position file:</td>
      <td class="table_text">[%user_bg_peak_pos_file%]</td>
    </tr>
  [%END%]

  <tr>
    <td class="table_title">Number of target sequences:</td>
    <td class="table_text">[%num_t_seqs%]</td>
  </tr>

  <tr>
    <td class="table_title">Number of background sequences:</td>
    <td class="table_text">[%num_bg_seqs%]</td>
  </tr>

  <tr>
    <td class="table_title">Total target sequence length:</td>
    <td class="table_text">[%t_seq_len%]</td>
  </tr>

  <tr>
    <td class="table_title">Total background sequence length:</td>
    <td class="table_text">[%bg_seq_len%]</td>
  </tr>

  <tr>
    <td class="table_title">Target sequence GC content:</td>
    <td class="table_text">
    [%IF t_seq_gc_content > high_seq_gc or t_seq_gc_content < low_seq_gc%]
      <b class="warn_table_text">[%t_seq_gc_content%]</b>
    [%ELSE%]
      [%t_seq_gc_content%]
    [%END%]
    </td>
  </tr>

  <tr>
    <td class="table_title">Background sequence GC content:</td>
    <td class="table_text">
    [%IF bg_seq_gc_content > high_seq_gc or bg_seq_gc_content < low_seq_gc%]
      <b class="warn_table_text">[%bg_seq_gc_content%]</b>
    [%ELSE%]
      [%bg_seq_gc_content%]
    [%END%]
    </td>
  </tr>

  [%IF tf_db%] 
    <tr>
      <td class="table_title">TFBS profile matrix source:</td>
      <td class="table_text">[%tf_db%]</td>
    </tr>
  [%END%]
  
  [%IF cl_db%] 
    <tr>
      <td class="table_title">TFBS cluster source:</td>
      <td class="table_text">[%cl_db%]</td>
    </tr>
  [%END%]
  
  [%IF collections%]
    <tr>
      <td class="table_title">JASPAR collection(s):</td>
      <td class="table_text">
        [%FOREACH collection = collections%][%collection%]&nbsp;[%END%]
      </td>
    </tr>
  [%END%]
    
  [%IF user_tf_family_file%]
    <tr>
      <td class="table_title">TF family file:</td>
      <td class="table_text">[%user_tf_family_file%]</td>
    </tr>
  [%END%]
  
  [%IF tf_families%]
    <tr>
      <td class="table_title">TFBS families:</td>
      <td class="table_text">
        [%FOREACH family = tf_families%][%family%]&nbsp;[%END%]
      </td>
    </tr>
  [%END%]
  
  [%IF tax_groups%]
    <tr>
      <td class="table_title">Taxonomic supergroup(s):</td>
      <td class="table_text">
        [%FOREACH tax_group = tax_groups%][%tax_group%]&nbsp;[%END%]
      </td>
    </tr>
  [%END%]
  
  [%IF min_ic%]
    <tr>
      <td class="table_title">Min. IC:</td>
      <td class="table_text">[%min_ic%] bits</td>
    </tr>
  [%END%]

  <tr>
    <td class="table_title">Matrix score threshold:</td>
    <td class="table_text">[%threshold%]</td>
  </tr>

  [%IF result_sort_by == 'zscore'%]
    [%display_sort_by = 'Z-score'%]
  [%ELSIF result_sort_by == 'fisher_p_value'%]
    [%display_sort_by = 'Fisher score'%]
  [%ELSIF result_sort_by == 'ks_p_value'%]
    [%display_sort_by = 'KS score'%]
  [%ELSE%]
    [%display_sort_by = result_sort_by%]
  [%END%]
  
  <tr>
    <td class="table_title">Results returned:</td>
    [%IF result_type == 'top_x_results'%]
      [%IF num_display_results == 'All'%]
        <td class="table_text">All results sorted by [%display_sort_by%]</td>
      [%ELSE%]
        <td class="table_text">Top [%num_display_results%] results sorted by [%display_sort_by%]</td>
      [%END%]
    [%ELSIF result_type == 'significant_hits'%]
      <td class="table_text">All results with a Z-score >= [%zscore_cutoff%] / Fisher score >= [%fisher_cutoff%] / KS p-value <= [%ks_cutoff%] sorted by [%display_sort_by%]</td>
    [%END%]
  </tr>

  <tr>
    <td class="table_title">The results will be sent to the following email:</td>
    <td class="table_text">[%email%]</td>
  </tr>
  
</table>


<h3>Results - click column headers to sort by that column</h3>

[%IF message%]
  <p class="warn_text">[%message%]</p>
[%ELSIF results%]
  [%IF warn_zero_bg_gene_hits%]
    <p class="warn_text"><b>Warning</b>: the provided background sequence set resulted in zero binding sites for one or more TFs. In these cases the Z-score is undefined and is marked as NA. In cases where there were one or more binding sites in the target gene set, the result could still be considered significant. Results for which this was the case are marked with an asterisk (*).
    </p>
  [%END%]

  <table id="results_table" class="sortable" cellpadding="2" cellspacing="0">
  <tr class="table_title">
    <th title="TFBS cluster name (click to sort)">TFBS cluster name</th>
    <th title="Target sequence hits (click to sort)">Target seq. hits</th>
    <th title="Target sequence non-hits (click to sort)">Target seq. non-hits</th>  
    <th title="Background sequence hits (click to sort)">Background seq. hits</th>  
    <th title="Background sequence non-hits (click to sort)">Background seq. non-hits</th>
    <th title="Target TFBS cluster hits (click to sort)">Target cluster hits</th>   
    <th title="Target TFBS cluster nucleotide rate (click to sort)">Target cluster nucleotide rate</th>
    <th title="Background TFBS cluster hits (click to sort)">Background cluster hits</th>
    <th title="Background TFBS cluster nucleotide rate (click to sort)">Background cluster nucleotide rate</th>
    <th title="Assesses TFBS clusters per basepair (click to sort)">Z-score</th>
    <th title="Assesses TFBS clusters per sequence (click to sort)">Fisher score</th>
    <th title="Assesses TFBS distances from sequence peaks (click to sort)">KS score</th>
  </tr>
  [%FOREACH result = results%]
    [%cl = tf_cluster_set.get_tf_cluster(result.id)%]
    <tr class="table_text">
      [%IF result.bg_gene_hits == 0 and result.t_gene_hits > 0%]
        <td><b class="warn_table_text"><a href="[%rel_htdocs_data_path%]/tfbs_clusters/c[%cl.id%]_info.html" target="_blank">*[%cl.name%]</a></b></td>
      [%ELSE%]
	    <td><a href="[%rel_htdocs_data_path%]/tfbs_clusters/c[%cl.id%]_info.html" target="_blank">[%cl.name%]</a></td>
      [%END%]

      <td align=right>
        [%IF result.t_gene_hits > 0%]
          <a href="[%rel_results_dir%]/c[%cl.id%].html" target="_blank">[%result.t_gene_hits%]</a>
        [%ELSE%]
          [%result.t_gene_hits%]
        [%END%]
      </td>

      <td align=right>[%result.t_gene_no_hits%]</td>

      <td align=right>
        [%IF result.bg_gene_hits == 0%]
          <b class="warn_table_text">[%result.bg_gene_hits%]</b>
        [%ELSE%]
          [%result.bg_gene_hits%]
        [%END%]
      </td>

      <td align=right>[%result.bg_gene_no_hits%]</td>

      <td align=right>
        [%IF result.t_cluster_hits > 0%]
          <a href="[%rel_results_dir%]/c[%cl.id%].html" target="_blank">[%result.t_cluster_hits%]</a>
        [%ELSE%]
          [%result.t_cluster_hits%]
        [%END%]
      </td>

      <td align=right>[%formatg(3, result.t_cluster_rate)%]</td>

      <td align=right>
        [%IF result.bg_gene_hits == 0%]
          <b class="warn_table_text">[%result.bg_cluster_hits%]</b>
        [%ELSE%]
          [%result.bg_cluster_hits%]
        [%END%]
      </td>

      <td align=right>
        [%IF result.bg_gene_hits == 0%]
          <b class="warn_table_text">[%formatf(3, result.bg_cluster_rate)%]</b>
        [%ELSE%]
          [%formatf(3, result.bg_cluster_rate)%]
        [%END%]
      </td>

      <td align=right>
        [%IF result.bg_gene_hits == 0%]
          <b class="warn_table_text">NA</b>
        [%ELSE%]
          [%formatf(3, result.zscore)%]
        [%END%]
      </td>

      [%IF result.fisher_p_value == 'inf' or result.fisher_p_value == 'Inf'%]
        <td sorttable_customkey="999999999" align=right nowrap>Inf</td>
      [%ELSE%]
        <td align=right nowrap>[%formatf(3, result.fisher_p_value)%]</td>
      [%END%]

      [%IF result.ks_p_value == 'inf' or result.ks_p_value == 'Inf'%]
        <td sorttable_customkey="999999999" align=right nowrap>Inf</td>
      [%ELSE%]
        <td align=right nowrap>[%formatf(3, result.ks_p_value)%]</td>
      [%END%]

<!--      
      <td align=right>
        [%IF result.ks_p_value == 'NA'%]
          <b class="warn_table_text">NA</b>
        [%ELSE%]
          [%formatf(3, result.ks_p_value)%]
        [%END%]
      </td>
-->

    </tr>
  [%END%]
  </table>

  <p class="text" align="center">
  <a href="[%rel_results_dir%]/[%results_file%]" target="_blank">Download as a tab delimited text file</a> (results will be kept on the server for [%result_retain_days%] days after analysis)
  </p>

  <p align="center"><a href="#top"><img src="[%rel_htdocs_path%]/images/arrow-up.gif" border="0" alt="Back to top"></a></p>

  <p class="text">
  Consult the <a href="javascript:newWindow=window.open('[%rel_htdocs_path%]/help.html','help','width=600,height=800,toolbar=1,location=0,directories=0,status=0,menuBar=0,scrollBars=1,resizable=0'); newWindow.focus()">help file</a> for an explanation of the column headings.
  </p>

[%ELSE%]
  <p class="warn_text">No TFBS clusters were found which matched your search criteria</p>
[%END%]
