<h3>Selected Parameters</h3>
<table border="0" cellpadding="2" cellspacing="0">
  <tr>
    <td class="table_title">Job ID:</td>
    <td class="table_text">[%job_id%]</td>
  </tr>
  <tr>
    <td class="table_title">Number of genes submitted for analysis:</td>
    <td class="table_text">[%num_t_gene_ids%]</td>
  </tr>
  <tr>
    <td class="table_title">Number of genes included in analysis:</td>
    <td class="table_text">[%num_t_included_gene_ids%]</td>
  </tr>
  <tr>
    <td class="table_title">Number of genes missing from analysis:</td>
    <td class="table_text">[%num_t_missing_gene_ids%]</td>
  </tr>
  <tr>
    <td class="table_title">Number of unique oPOSSUM genes in analysis:</td>
    <td class="table_text">[%num_t_gids%]</td>
  </tr>
  <tr>
    <td class="table_title">JASPAR collection(s):</td>
    [%using_core_collection = 0%]
    <td class="table_text">
      [%FOREACH collection = collections%]
        [%IF collection == 'CORE'%]
          [%using_core_collection = 1%]
        [%END%]
        [%collection%]&nbsp;
      [%END%]
    </td>
  </tr>
  [%IF using_core_collection == 1%]
    [%IF tax_groups%]
      <tr>
        <td class="table_title">Taxonomic supergroup(s):</td>
        <td class="table_text">
          [%FOREACH tax_group = tax_groups%]
            [%tax_group%]&nbsp;
          [%END%]
        </td>
      </tr>
    [%END%]
  [%END%]
  <tr>
    <td class="table_title">Anchoring TF:</td>
    <td class="table_text">[%anchor_matrix.name()%]</td>
  </tr>
  <tr>
    <td class="table_title">Maximum Inter-Binding Distance:</td>
    <td class="table_text">[%max_site_distance%]</td>
  </tr>

  [%IF tf_select_criteria == 'min_ic'%]
    <tr>
      <td class="table_title">Minimum profile specificity:</td>
      <td class="table_text">[%min_ic%] bits</td>
    </tr>
  [%ELSIF tf_select_criteria == 'specific'%]
    [%IF tf_set%]
      [%matrix_list = tf_set.get_matrix_list('name')%]
      <tr>
        <td class="table_title">TFBS profiles:</td>
        <td class="table_text">
          [%FOREACH matrix = matrix_list%]
            [%matrix.name()%] 
          [%END%]
        </td>
      </tr>
    [%END%]
  [%END%]
  <tr>
    <td class="table_title">Conservation cutoff:</td>
    <td class="table_text">[%min_conservation%]</td>
  </tr>
  <tr>
    <td class="table_title">Matrix score threshold:</td>
    <td class="table_text">[%threshold*100%]%</td>
  </tr>
  <tr>
    <td class="table_title">Upstream sequence length:</td>
    <td class="table_text">[%upstream_bp%]</td>
  </tr>
  <tr>
    <td class="table_title">Downstream sequence length:</td>
    <td class="table_text">[%downstream_bp%]</td>
  </tr>
<!--
  <tr>
    <td class="table_title">Target GC content:</td>
    <td class="table_text">
    [%IF t_cr_gc_content > high_seq_gc or t_cr_gc_content < low_seq_gc%]
      <b class="warn_text">[%t_cr_gc_content%]</b>
    [%ELSE%]
      [%t_cr_gc_content%]
    [%END%]
    </td>
  </tr>
  <tr>
    <td class="table_title">Background GC content:</td>
    <td class="table_text">
    [%IF bg_cr_gc_content > high_seq_gc or bg_cr_gc_content < low_seq_gc%]
      <b class="warn_text">[%bg_cr_gc_content%]</b>
    [%ELSE%]
      [%bg_cr_gc_content%]
    [%END%]
    </td>
  </tr>
-->

  [%IF result_sort_by == 'zscore'%]
    [%display_sort_by = 'Z-score'%]
  [%ELSIF result_sort_by == 'fisher_p_value'%]
    [%display_sort_by = 'Fisher score'%]
  [%ELSE%]
    [%display_sort_by = result_sort_by%]
  [%END%]
  <tr>
    <td class="table_title">Results returned:</td>
    [%IF result_type == 'top_x_results'%]
      [%IF num_display_results == 'All'%]
        <td class="table_text">All results sorted by [%display_sort_by%]</td>
      [%ELSE%]
        <td class="table_text">Top [%num_display_results%] results sorted by [%display_sort_by%]</td>
      [%END%]
    [%ELSIF result_type == 'significant_hits'%]
      <td class="table_text">All results with a Z-score >= [%zscore_cutoff%] and a Fisher score >= [%fisher_cutoff%] sorted by [%display_sort_by%]</td>
    [%END%]
  </tr>
</table>

<h3>Target Genes</h3>
<table border="0" cellpadding="2" cellspacing="0">
  <tr>
    <td class="table_title" valign="top"><a href="javascript:newWindow=window.open('[%rel_htdocs_path%]/help.html#genes_analyzed','help','width=600,height=800,toolbar=1,location=0,directories=0,status=0,menuBar=0,scrollBars=1,resizable=0'); newWindow.focus()">Analyzed:</a></td>
    <td class="table_text">[%FOREACH gene_id = t_included_gene_ids%] [%gene_id%][%END%]</td>
  </tr>

  [%IF num_t_missing_gene_ids > 0%]
  <tr>
    <td class="table_title" valign="top"><a href="javascript:newWindow=window.open('[%rel_htdocs_path%]/help.html#genes_excluded','help','width=600,height=800,toolbar=1,location=0,directories=0,status=0,menuBar=0,scrollBars=1,resizable=0'); newWindow.focus()">Omitted:</a></td>
    <td class="table_text">
      [%FOREACH gene_id = t_missing_gene_ids%] [%gene_id%][%END%]
    </td>
  </tr>
  [%END%]
</table>
</p>

<!--
[%IF analysis_type == 'custom' and bg_id_input_method != 'all'%]
  <h3>Background Genes</h3>
  <table border="0" cellpadding="2" cellspacing="0">
    <tr>
      <td class="table_title"><a href="javascript:newWindow=window.open('[%rel_htdocs_path%]/help.html#bg_genes_analyzed','help','width=600,height=800,toolbar=1,location=0,directories=0,status=0,menuBar=0,scrollBars=1,resizable=0'); newWindow.focus()">Analyzed:</a></td>
<td class="table_text">[%FOREACH gene_id = bg_gene_ids%] [%gene_id%][%END%]</td>
    </tr>
    [%IF missing_bg_gene_ids.0%]
    <tr>
      <td class="table_title"><a href="javascript:newWindow=window.open('[%rel_htdocs_path%]/help.html#bg_genes_excluded','help','width=600,height=800,toolbar=1,location=0,directories=0,status=0,menuBar=0,scrollBars=1,resizable=0'); newWindow.focus()">Omitted:</a></td>
      <td class="table_text">
        [%FOREACH gene_id = missing_bg_gene_ids%] [%gene_id%][%END%]
      </td>
    </tr>
    [%END%]
  </table>
  </p>
[%END%]
-->

<h3>Results - click column headers to sort by that column</h3>
[%IF message%]
  <p class="warn_text">[%message%]</p>
[%ELSIF results%]
  [%IF warn_zero_bg_hits%]
    <p class="warn_text"><b>Warning</b>: The provided background set resulted in zero binding sites for one or more TFs. These cases are highlighted in red. In cases where there were also zero binding sites for the target set, the Z-score is undefined (NA). In cases where there were one or more binding sites in the target set, the Z-score is infinite (Inf). However, the result may or may not be considered significant. Results for which this was the case are marked with an asterisk (*).
    </p>
  [%END%]

  <table id="results_summary_table" class="sortable" cellpadding="2" cellspacing="0">
  <tr class="table_title">
    <th title="TF name (click to sort)">TF</th>
    <th title="JASPAR ID (click to sort)">JASPAR ID</th>
    <th title="TF class (click to sort)">Class</th>
    <th title="TF family (click to sort)">Family</th>
    <th title="TF taxonomic supergroup (click to sort)">Tax group</th>
    <th title="TF information content (click to sort)">IC</th>
    <th title="TF GC content (click to sort)">GC Content</th>
    <th title="Target gene hits (click to sort)">Target gene hits</th>
    <th title="Target gene non-hits (click to sort)">Target gene non-hits</th>
    <th title="Background gene hits (click to sort)">Background gene hits</th>
    <th title="Background gene non-hits (click to sort)">Background gene non-hits</th>
    <th title="Target TFBS hits (click to sort)">Target TFBS hits</th>
    <th title="Target TFBS nucleotide rate (click to sort)">Target TFBS nucleotide rate</th>
    <th title="Background TFBS hits (click to sort)">Background TFBS hits</th>
    <th title="Background TFBS nucleotide rate (click to sort)">Background TFBS nucleotide rate</th>
    <th title="Assesses the enrichment of TFBS bp density (click to sort)">Z-score</th>
    <th title="Assesses the enrichment of genes with one or more TFBS (click to sort)">Fisher score</th>
  </tr>

  [%FOREACH result = results%]
    [%tf = tf_set.get_tf(result.id)%]
    <tr class="table_text">
      [%IF result.bg_gene_hits == 0 and result.t_gene_hits > 0%]
        <td><b class="warn_table_text">*[%tf.name%]</b></td>
      [%ELSE%]
        <td>[%tf.name%]</td>
      [%END%]
      [%collection = tf.tag('collection')%]
      <td>
      [%IF collection != 'PENDING'%]
        <a href="javascript:newWindow=window.open('[%jaspar_url%]?rm=present&collection=[%collection%]&ID=[%tf.ID%]', '[%tf.ID%]', 'width=850,height=700,toolbar=0,location=0,directories=0,status=0,menuBar=0,scrollBars=1'); newWindow.focus()">[%tf.ID%]</a>
      [%ELSE%]
        [%tf.ID%]
      [%END%]
      </td>
      <td>[%tf.class%]</td>
      <td>[%tf.tag('family')%]</td>
      <td>[%tf.tag('tax_group')%]</td>
      [%total_ic = formatf(3, tf.to_ICM.total_ic)%]
      <td align=right>
      [%IF total_ic > high_matrix_ic or total_ic < low_matrix_ic%]
        <b class="warn_table_text">[%total_ic%]</b>
      [%ELSE%]
        [%total_ic%]
      [%END%]
      </td>
      [%gc = tf.tag('gc_content')%]
      <td align=right>
      [%IF gc > high_matrix_gc or gc < low_matrix_gc%]
        <b class="warn_table_text">[%gc%]</b>
      [%ELSE%]
        [%gc%]
      [%END%]
      </td>
      <td align=right>
        [%IF result.t_gene_hits > 0%]
          <a href="[%rel_results_dir%]/[%tf.ID%].html" target="_blank">[%result.t_gene_hits%]</a>
        [%ELSE%]
          [%result.t_gene_hits%]
        [%END%]
      </td>
      <td align=right>[%result.t_gene_no_hits%]</td>
      <td align=right>
      [%IF result.bg_gene_hits == 0%]
        <b class="warn_table_text">[%result.bg_gene_hits%]</b>
      [%ELSE%]
        [%result.bg_gene_hits%]
      [%END%]
      </td>
      <td align=right>[%result.bg_gene_no_hits%]</td>
      <td align=right>
        [%IF result.t_tfbs_hits > 0%]
          <a href="[%rel_results_dir%]/[%tf.ID%].html" target="_blank">[%result.t_tfbs_hits%]</a>
        [%ELSE%]
          [%result.t_tfbs_hits%]
        [%END%]
      </td>
      <td align=right>[%formatg(3, result.t_tfbs_rate)%]</td>
      <td align=right>
      [%IF result.bg_gene_hits == 0%]
        <b class="warn_table_text">[%result.bg_tfbs_hits%]</b>
      [%ELSE%]
        [%result.bg_tfbs_hits%]
      [%END%]
      </td>
      <td align=right>
      [%IF result.bg_gene_hits == 0%]
        <b class="warn_table_text">[%formatg(3, result.bg_tfbs_rate)%]</b>
      [%ELSE%]
        [%formatg(3, result.bg_tfbs_rate)%]
      [%END%]
      </td>

      [%IF result.bg_gene_hits == 0%]
        [%IF result.zscore == 'inf' or result.zscore == 'Inf'%]
          <td sorttable_customkey="999999999" align=right nowrap><b class="warn_table_text">Inf</b></td>
        [%ELSE%]
          <td align=right nowrap><b class="warn_table_text">NA</b></td>
        [%END%]
      [%ELSE%]
        <td align=right nowrap>[%formatf(3, result.zscore)%]</td>
      [%END%]

      [%IF result.fisher_p_value == 'inf' or result.fisher_p_value == 'Inf'%]
        <td sorttable_customkey="999999999" align=right nowrap>Inf</td>
      [%ELSE%]
        <td align=right nowrap>[%formatf(3, result.fisher_p_value)%]</td>
      [%END%]
    </tr>
  [%END%]
  </table>
  <p class="text" align="center">
  <a href="[%rel_results_dir%]/[%results_file%]" target="_blank">Download as a tab delimited text file</a> (results will be kept on the server for [%result_retain_days%] days after analysis)
  </p>

  <p class="text" align="center">
  View plot of <a href="[%rel_results_dir%]/[%zscore_plot_file%]" target="_blank">Z-score vs. %GC content</a>
  <br>
  View plot of <a href="[%rel_results_dir%]/[%fisher_plot_file%]" target="_blank">Fisher score vs. %GC content</a>
  </p>

  <p align="center"><a href="#top"><img src="[%rel_htdocs_path%]/images/arrow-up.gif" border="0" alt="Back to top"></a></p>

  <p class="text">
  Consult the <a href="javascript:newWindow=window.open('[%rel_htdocs_path%]/help.html','help','width=600,height=800,toolbar=1\
  ,location=0,directories=0,status=0,menuBar=0,scrollBars=1,resizable=0'); newWindow.focus()">help file</a> for an explanation of the column headings.
  </p>
[%ELSE%]
  <p class="warn_text">No TFBSs were found which matched your search criteria</p>
[%END%]
